### PR TITLE
BUG: test, fix regression in converting to ctypes

### LINKED ISSUE #14214: ctypeslib.as_ctypes cannot wrap large arrays (> 2 ** 31 bytes)
After updating Numpy I can no longer use `numpy.ctypeslib.as_ctypes` for an array whose size is larger than 2 ** 31 bytes (a signed C int). This appears to be due to the changes from 08a92de354f1f7592c7957f4aafe892f3d5a828a that allow more general ctypes pointers.

Old style pointer:
https://github.com/numpy/numpy/blob/971e2e89d08deeae0139d3011d15646fdac13c92/numpy/ctypeslib.py#L402-L404

New style pointer:
https://github.com/numpy/numpy/blob/5a5987291dc95376bb098be8d8e5391e89e77a2c/numpy/ctypeslib.py#L538-L540

I'm using ctypeslib in the context of making large `multiprocessing.sharedctypes.Array` objects available to multiprocessing pools. 

I would wager that this C int size limit runs deep, making a direct fix difficult. I could simply monkey-patch to previous functionality, since I know that my arrays work with the previous, more restrictive wrapping. But I think splitting large (> 2 GB) arrays across pool workers is probably a common use case (I hope I didn't miss an existing issue). 

I'm a bit out of depth, but would it make sense to fall back to the old behavior after testing for no-strides and native endian-ness?

### Reproducing code example:

```python
import numpy as np
arr = np.zeros((2 ** 31 + 1,), 'b')
assert(arr.size * arr.itemsize > 2 ** 31)
c_arr = np.ctypeslib.as_ctypes(arr)      
```

Under versions > 1.16.0 I am getting an error (traceback below), because the dtype memory footprint doesn't fit into a signed C int. 

### Error message:

```python                                                                                                            
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-4-e91c3f6f3634> in <module>
----> 1 c_arr = np.ctypeslib.as_ctypes(arr)

~/.pyenv/versions/3.6.4/envs/ctype_test/lib/python3.6/site-packages/numpy/ctypeslib.py in as_ctypes(obj)
    536             raise TypeError("readonly arrays unsupported")
    537 
--> 538         dtype = _dtype((ai["typestr"], ai["shape"]))
    539         result = as_ctypes_type(dtype).from_address(addr)
    540         result.__keep = obj

ValueError: invalid shape in fixed-type tuple: dtype size in bytes must fit into a C int.
```

### Numpy/Python version information:
I down-versioned numpy until `as_ctypes` worked again:

Result of `import sys, numpy; print(numpy.__version__, sys.version)`

*Where it works*
```
1.16.0 3.6.4 (default, Feb 22 2018, 18:37:35) 
[GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.39.2)]
```

*Where it breaks*
```
1.17.0 3.6.4 (default, Feb 22 2018, 18:37:35) 
[GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.39.2)]
```

### PR DESCRIPTION
**Pull request by mattip**
Fixes gh-14214, using the test and code in the issue.

### CODE REVIEW

**Comment by mattip** [comment]
64bit windows chokes on `2147483649 * ctypes.c_byte`, and 32 bit systems don't like `arr = np.zeros((2 ** 31 + 1,), 'b')`

File: numpy/core/tests/test_regression.py [hunk 1]
```diff
@@ -2500,3 +2500,12 @@ def __len__(self):
         t = T()
         #gh-13659, would raise in broadcasting [x=t for x in result]
         np.array([t])
+
+    @pytest.mark.skipif(sys.maxsize < 2 ** 31 + 1, reason='overflows 32-bit python')
+    @pytest.mark.skipif(sys.platform == 'win32', reason='overflows on windows')
+    def test_to_ctypes(self):
+        #gh-14214
+        arr = np.zeros((2 ** 31 + 1,), 'b')
+        assert(arr.size * arr.itemsize > 2 ** 31)
```
Comment (eric-wieser) [inline numpy/core/tests/test_regression.py]: nit: parens here are weird.

Could use our `assert_(...)`, but probably best to drop them as `assert ...` and let pytest handle the message
```suggestion
        assert arr.size * arr.itemsize > 2 ** 31
```

File: numpy/ctypeslib.py [hunk 2]
```diff
@@ -535,7 +535,10 @@ def as_ctypes(obj):
         if readonly:
             raise TypeError("readonly arrays unsupported")
 
-        dtype = _dtype((ai["typestr"], ai["shape"]))
-        result = as_ctypes_type(dtype).from_address(addr)
+        # can't use `_dtype((ai["typestr"], ai["shape"]))` here, as it overflows
+        # dtype.itemsize (gh-14214)
+        ctype_scalar = as_ctypes_type(ai["typestr"])
+        result_type = _ctype_ndarray(ctype_scalar , ai['shape'])
```
Comment (eric-wieser) [inline numpy/ctypeslib.py]:
```suggestion
        result_type = _ctype_ndarray(ctype_scalar, ai["shape"])
```

File: numpy/core/tests/test_regression.py [hunk 3]
```diff
@@ -2500,3 +2500,12 @@ def __len__(self):
         t = T()
         #gh-13659, would raise in broadcasting [x=t for x in result]
         np.array([t])
+
+    @pytest.mark.skipif(sys.maxsize < 2 ** 31 + 1, reason='overflows 32-bit python')
+    @pytest.mark.skipif(sys.platform == 'win32', reason='overflows on windows')
```
Comment (eric-wieser) [inline numpy/core/tests/test_regression.py]: Is this second check necessary? It doesn't make sense to me why this would overflow.

Comment (eric-wieser) [reply numpy/core/tests/test_regression.py]: Looks like it was fixed in https://github.com/python/cpython/pull/3006. Can we combine this condition with `sys.version_info[:1] < 3.8` and add a reference to bpo-16865?

**Review by eric-wieser — APPROVED** [review]

**Comment by charris** [comment]
Thanks Matti.

**Pull request merged**